src/ray/core_worker/store_provider/store_provider.h

```diff
@@ -1,36 +1,72 @@
 #ifndef RAY_CORE_WORKER_STORE_PROVIDER_H
 #define RAY_CORE_WORKER_STORE_PROVIDER_H
 
+#include "arrow/python/serialize.h"
+
 #include "ray/common/buffer.h"
 #include "ray/common/id.h"
 #include "ray/common/status.h"
 #include "ray/core_worker/common.h"
 
 namespace ray {
 
-/// Binary representation of ray object.
+/// RayObjects consist of data and metadata. Metadata is always passed
+/// in as a buffer because it is small, but data can be handled
+/// differently by implementations in order to avoid copying large data.
 class RayObject {
  public:
-  /// Create a ray object instance.
-  ///
-  /// \param[in] data Data of the ray object.
-  /// \param[in] metadata Metadata of the ray object.
-  RayObject(const std::shared_ptr<Buffer> &data, const std::shared_ptr<Buffer> &metadata)
-      : data_(data), metadata_(metadata) {}
+  RayObject(const std::shared_ptr<Buffer> &metadata)
+      : metadata_(metadata) {}
+
+  virtual ~RayObject() {}
 
-  /// Return the data of the ray object.
-  const std::shared_ptr<Buffer> &GetData() const { return data_; };
```

Comment:
I'm feeling we probably need to have a virtual function for `GetData()`, otherwise to retrieve the data for a `RayObject`, user has to allocate a buffer and copy the data there using `WriteDataTo`, which sounds a bit too heavy. Or user has to use downcast which is not recommended, or sometimes is not always possible.

Reply:
Yeah I thought about this too - the reason that I didn't keep it is that `GetData()` will incur a copy for the `PyArrowObject` and we don't actually use `GetData()` anywhere except in the tests.

I'm fine with either punting this until if/when we actually use it or including it with a big disclaimer for the `PyArrowObject`. What do you think is better?

Reply:
Ah, I see now that this is used in the Java code. I'll update to include a `GetData()` method.

Reply:
yep, thanks